README
This is dcload-ip 1.0.3 with various fixes/changes by Tim Hentenaar <tim@hentenaar.com>.

You can find the latest version at: http://github.com/thentenaar/dc-load-ip

Here is a brief list of changes from the original:

	* Added some fixes to fix compiler errors/warnings
	* Added maple passthrough functionality

Makefile Changes

	I've made various changes to the Makefiles according to my setup. 
	For instance, it's assumed that the 'scramble' utility will be
	in the 'make-cd' directory when making a CD image.

	You'll want to review Makefile.cfg and edit it according to your
	needs.

Maple Passthrough

	I've added a feature which I like to call 'Maple Passthrough.' 
	Basically, this allows the portion of dcload-ip that runs on the
	Dreamcast to act as a bridge between an application on the network
	and the maple bus.

	The way this works is fairly simple. You send a UDP datagram to 
	port 31313, containing the following:

	'MAPL' (4 byte command ID)
	Port   (8-bit port number)
	Slot   (8-bit slot number)
	Cmd    (8-bit Maple command ID)
	Last   (8-bit Maple last flag) [Ignored]
	Length (8-bit Maple data length (in 32-bit words))
	Data   (32-bit words: Length, [data]) specified if Length > 0

	The response packet is as follows:
	
	Status (8-bit Maple status)
	To     (8-bit Maple address)
	From   (8-bit Maple address)
	Length (8-bit Maple data length (in 32-bit words))
	Data   (32-bit words)

	*** See Maple.pm for a Perl module illustrating how to use this functionality. ***

Note that this software isn't perfect, and that with all the changes to GCC over the years since the last revision to
the original dcload-ip, some features (e.g. exception handling) may be a bit buggy or may not work at all.

The following is the original README:

dcload-ip 1.0.3 - a Dreamcast ethernet loader by <andrewk@napalm-x.com>

Features

* Load elf, srec, and bin
* PC I/O (read, write, etc to PC - compatible with original dcload)
* Exception handler
* Extremely fast (at least for me - I get 1130Kbyte/sec on 10mbit half-duplex)
* Now works on 100mbit (I get 2580Kbyte/sec on 100mbit half-duplex)

Building

1. Edit Makefile.cfg for your system and network, and then run make

Installation

1. PC - run make install (installs dc-tool)
2. DC
   a. cd make-cd, edit Makefile, insert blank cd-r, run make
 or 
   b. take target-src/1st_read/1st_read.bin and stuff it on a cd yourself
      (please use the IP.BIN from the make-cd directory if you are going
       to distribute either cds or cd images)
  
On-screen display

* If you see the message "NO BROADBAND ADAPTER DETECTED!" or 
  "ERROR INITIALIZING BROADBAND ADAPTER", something has gone wrong.
* The correct display is something like:
  dcload-ip 1.0.3        <- name/version
  00:d0:f1:02:ab:dd      <- dc hardware address
  c0.a8.00.04            <- dc ip address (in hex)
  idle...                <- status

* If the status line reports "link change..." and does not change back to 
  "idle..." within a short period of time, you may have a cable problem.
  dcload-ip will not work while "link change..." is displayed. The 
  "link change..." message normally is seen when you start dcload-ip, when
  you execute dc-tool -r, and when you disconnect the ethernet cable.

Testing

1. cd example-src
2. dc-tool -x console-test (tests some PC I/O)
3. dc-tool -x exception-test (generates an exception)
4. dc-tool -x gethostinfo (displays the Dreamcast's ip, and the ip and port of
   the dc-tool host)

Notes

* You can use arp instead of setting the dreamcast's ip in Makefile.cfg
* Tested systems: Debian GNU/Linux 2.2
* There are almost certainly bugs
* Patches and improvements are welcome

Credits
* rtl8139 code based on code by Dan Potter
* There are some various files from newlib-1.8.2 here
* video.s was written by Marcus Comstedt
* initial win32 porting and implementation of -t by Florian 'Proff' Schulze
* win32 bugfix by The Gypsy
* fixes for cygwin by Florian 'Proff' Schulze
* rx config bug pointed out by Slant
* horridly buggy nature of 1.0.1 + 1.0.2 pointed out by Dan Potter


CHANGES
WHATS NEW IN 1.0.3

* actually seems to be working right, unlike 1.0.1 and 1.0.2, which were both
  buggy as hell under many (most?) conditions. If 1.0.1 and 1.0.2 locked up, 
  went berzerk, or did other strange and horrible things when you used them,
  try 1.0.3.
* dc-tool should now be able to recover from dropped packets in nearly all
  cases (prior versions would just get stuck if certain packets got dropped)
* gethostinfo syscall now returns the dreamcast's ip, in addition to the ip
  and port of the dc-tool host
* dcload-ip now reports "link change..." on the status line while the rtl8139
  is negotiating. This message should be displayed momentarily when dcload-ip
  starts, when you execute dc-tool -r, or if you disconnect and reconnect the
  ethernet cable. If the message stays there, check your ethernet cable.

WHATS NEW IN 1.0.2

* fixed bug where rx configuration was not restored properly
  (only affected dcload syscalls - could cause dc-tool to segfault apparently,
   and who knows what else...)

WHATS NEW IN 1.0.1

* works with 100mbit now
* added syscall to return the ip and port of the dc-tool host
* added O_BINARY to all open()s in dc-tool to help cygwin users

WHATS NEW IN 1.0.0

* initial release of dcload-ip



Maple.pm
#!/usr/bin/perl -w
# vi:set ts=4:

package Maple;
use strict;

BEGIN {
	use Exporter;
	use Socket;

	our @ISA    = qw(Exporter);
	our @EXPORT = qw(MAPLE_CMD_DEVINFO MAPLE_CMD_ALLINFO MAPLE_CMD_RESET 
	                 MAPLE_CMD_SHUTDOWN MAPLE_RESP_DEVINFO MAPLE_RESP_DEVINFOEX
	                 MAPLE_RESP_ACK MAPLE_RESP_DATAXFER MAPLE_CMD_GETCOND
	                 MAPLE_CMD_GETMEMINFO MAPLE_CMD_READBLOCK MAPLE_CMD_WRITEBLOCK 
	                 MAPLE_CMD_SYNCBLOCK MAPLE_CMD_SETCOND MAPLE_CMD_MICROPHONE
	                 MAPLE_CMD_CAMERA MAPLE_RESP_ERR MAPLE_RESP_BADFUNC
	                 MAPLE_RESP_BADCMD MAPLE_RESP_AGAIN MAPLE_RESP_FILEERR
	                 MAPLE_FUNC_CONTROLLER MAPLE_FUNC_MEMORY MAPLE_FUNC_LCD
	                 MAPLE_FUNC_CLOCK MAPLE_FUNC_MICROPHONE MAPLE_FUNC_AR_GUN
	                 MAPLE_FUNC_KEYBOARD MAPLE_FUNC_LIGHT_GUN MAPLE_FUNC_PURU_PURU
	                 MAPLE_FUNC_MOUSE MAPLE_FUNC_CAMERA send_maple_packet);
}

=head1 NAME

Maple - (Very) Simple Dreamcast Maple-over-UDP Client

=head1 DESCRIPTION

C<Maple> provides a function to send Maple packets (and read the response) 
from a Dreamcast running a modified version of dcload-ip, available from
L<http://github.com/thentenaar/dc-load-ip>.

A summary of the maple protocol is available at: 
L<http://mc.pp.se/dc/maplebus.html>

=head1 SYNOPSIS

	use Maple;

	my %cmd  = (
		'cmd'    => MAPLE_CMD_RESET,
		'port'   => 0,
		'slot'   => 1,
		'last'   => 0,
		'length' => 0
	);

	my $resp = send_maple_packet('127.0.0.1',\%cmd);

This example sends a B<Reset> command to the device on port 0, slot 1 of the
Dreamcast with the IP B<127.0.0.1>.

	use Maple;

	my %cmd  = (
		'cmd'    => MAPLE_CMD_GETMEMINFO,
		'port'   => 0,
		'slot'   => 1,
		'last'   => 0,
		'length' => 2,
		'data'   => [ MAPLE_FUNC_MEMORY, 0 ]
	);

	my $resp = send_maple_packet('127.0.0.1',\%cmd);

This example retrieves the memory info for partition 0 of the block
device at port 0, slot 1. The C<data> parameter should be specified
as an array of 32-bit unsigned integers.

The B<MAPLE_CMD_*>, B<MAPLE_RESP_*>, and B<MAPLE_FUNC_*> constants can
be found within this module.

=cut

# Maple Commands/Responses
use constant MAPLE_CMD_DEVINFO     => 0x01;
use constant MAPLE_CMD_ALLINFO     => 0x02;
use constant MAPLE_CMD_RESET       => 0x03;
use constant MAPLE_CMD_SHUTDOWN    => 0x04;
use constant MAPLE_RESP_DEVINFO    => 0x05;
use constant MAPLE_RESP_DEVINFOEX  => 0x06;
use constant MAPLE_RESP_ACK        => 0x07;
use constant MAPLE_RESP_DATAXFER   => 0x08;
use constant MAPLE_CMD_GETCOND     => 0x09;
use constant MAPLE_CMD_GETMEMINFO  => 0x0a;
use constant MAPLE_CMD_READBLOCK   => 0x0b;
use constant MAPLE_CMD_WRITEBLOCK  => 0x0c;
use constant MAPLE_CMD_SYNCBLOCK   => 0x0d;
use constant MAPLE_CMD_SETCOND     => 0x0e;
use constant MAPLE_CMD_MICROPHONE  => 0x0f;
use constant MAPLE_CMD_CAMERA      => 0x11;
use constant MAPLE_RESP_ERR        => -1;
use constant MAPLE_RESP_BADFUNC    => -2;
use constant MAPLE_RESP_BADCMD     => -3;
use constant MAPLE_RESP_AGAIN      => -4;
use constant MAPLE_RESP_FILEERR    => -5;

# Maple Functions
use constant MAPLE_FUNC_CONTROLLER => 0x00000001;
use constant MAPLE_FUNC_MEMORY     => 0x00000002;
use constant MAPLE_FUNC_LCD        => 0x00000004;
use constant MAPLE_FUNC_CLOCK      => 0x00000008;
use constant MAPLE_FUNC_MICROPHONE => 0x00000010;
use constant MAPLE_FUNC_AR_GUN     => 0x00000020;
use constant MAPLE_FUNC_KEYBOARD   => 0x00000040;
use constant MAPLE_FUNC_LIGHT_GUN  => 0x00000080;
use constant MAPLE_FUNC_PURU_PURU  => 0x00000100;
use constant MAPLE_FUNC_MOUSE      => 0x00000200;
use constant MAPLE_FUNC_CAMERA     => 0x00000800;

=head1 FUNCTIONS

=head2 send_maple_packet(IP,MAPLE_DATA)

Sends a maple packet, and reads the response.

B<IP> is the IP address of the Dreamcast as a string.
B<MAPLE_DATA> is a hashref of values, as in the synopsis.

The response is returned as a hashref with the following fields:
    B<Status> - Maple status code (One of B<MAPLE_RESP_*>)
    B<To>     - Destination maple address
    B<From>   - Source maple address
    B<Length> - Length of data (number of 32-bit words)
    B<Data>   - Data (string of B<Length>*4 bytes)

B<undef> is returned if the B<bind> syscall fails.

=cut

sub send_maple_packet($$) {
	my ($ip,$maple) = @_;
	return undef unless (defined $ip && defined $maple);

	# Create/Bind the socket
	my $sin = sockaddr_in('31313',inet_aton($ip));
	my $slo = sockaddr_in('31314',inet_aton('0.0.0.0'));
	socket(MSOCK,PF_INET,SOCK_DGRAM,getprotobyname('udp'));
	bind(MSOCK,$slo) || return undef;

	# Build the packet
	my $bufi;
	my $data = 'MAPL' . 
	         pack('c5',$maple->{'port'},$maple->{'slot'},$maple->{'cmd'},$maple->{'last'},$maple->{'length'});
	$data .= pack('N' . int($maple->{'length'}),@{$maple->{'data'}}) if (defined($maple->{'data'}) && $maple->{'length'} > 0);

	# Send packet and get response
	send(MSOCK,$data,0,$sin);
	recv(MSOCK,$bufi,1024,0);
	close(MSOCK);

	# Parse Response
	my %resp = ();
	($resp{'status'},$resp{'to'},$resp{'from'},$resp{'length'}) = unpack('c4',$bufi);
	$resp{'data'} = substr($bufi,4,$resp{'length'} << 2) if ($resp{'status'} > 0 && $resp{'length'} > 0);
	return \%resp;
}

=head1 AUTHOR

Tim Hentenaar E<lt>tim@hentenaar.comE<gt>

=head1 VERSION

1.0 (Jun 23 2011)

=head1 COPYRIGHT

Copyright (C) 2011 Tim Hentenaar.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

=cut

1;
